### PR TITLE
MongoDB table engine now establishes connection only when it reads data.

### LINKED ISSUE #20055: ch start failed: Caught exception while loading metadata:
2021.02.04 10:53:13.119713 [ 27702 ] {} <Error> Application: Caught exception while loading metadata: Poco::Exception. Code: 1000, e.code() = 111, e.displayText() = Connection refused: 47.106.193.101:7017, Stack trace (when copying this message, always include the lines below):

0. Poco::IOException::IOException(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int) @ 0x13cda46c in /usr/bin/clickhouse
1. Poco::Net::ConnectionRefusedException::ConnectionRefusedException(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int) @ 0x13bf7299 in /usr/bin/clickhouse
2. ? @ 0xa1a7374 in /usr/bin/clickhouse
3. Poco::Net::SocketImpl::connect(Poco::Net::SocketAddress const&) @ 0x13c0ccb9 in /usr/bin/clickhouse
4. Poco::MongoDB::Connection::Connection(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int) @ 0x11b5f404 in /usr/bin/clickhouse
5. DB::StorageMongoDB::StorageMongoDB(DB::StorageID const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, unsigned short, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, DB::ColumnsDescription const&, DB::ConstraintsDescription const&, DB::Context const&) @ 0x1177c4d5 in /usr/bin/clickhouse
6. ? @ 0x1177d5af in /usr/bin/clickhouse
7. DB::StorageFactory::get(DB::ASTCreateQuery const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, DB::Context&, DB::Context&, DB::ColumnsDescription const&, DB::ConstraintsDescription const&, bool) const @ 0x10f1fbc6 in /usr/bin/clickhouse
8. DB::createTableFromAST(DB::ASTCreateQuery, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, DB::Context&, bool) @ 0x11516e5f in /usr/bin/clickhouse
9. ? @ 0x1151e6e1 in /usr/bin/clickhouse
10. ? @ 0x1151f02d in /usr/bin/clickhouse
11. ThreadPoolImpl<ThreadFromGlobalPool>::worker(std::__1::__list_iterator<ThreadFromGlobalPool, void*>) @ 0xa43d6ad in /usr/bin/clickhouse
12. ThreadFromGlobalPool::ThreadFromGlobalPool<void ThreadPoolImpl<ThreadFromGlobalPool>::scheduleImpl<void>(std::__1::function<void ()>, int, std::__1::optional<unsigned long>)::'lambda1'()>(void&&, void ThreadPoolImpl<ThreadFromGlobalPool>::scheduleImpl<void>(std::__1::function<void ()>, int, std::__1::optional<unsigned long>)::'lambda1'()&&...)::'lambda'()::operator()() const @ 0xa43dd93 in /usr/bin/clickhouse
13. ThreadPoolImpl<std::__1::thread>::worker(std::__1::__list_iterator<std::__1::thread, void*>) @ 0xa43cc4d in /usr/bin/clickhouse
14. ? @ 0xa43b3ff in /usr/bin/clickhouse
15. start_thread @ 0x7dd5 in /usr/lib64/libpthread-2.17.so
16. __clone @ 0xfdead in /usr/lib64/libc-2.17.so
 (version 20.8.3.18)
2021.02.04 10:53:13.119847 [ 27702 ] {} <Information> Application: Shutting down storages.
2021.02.04 10:53:13.119863 [ 27702 ] {} <Information> Context: Shutdown disk default
2021.02.04 10:53:13.119941 [ 27729 ] {} <Trace> SystemLog (system.query_log): Terminating
2021.02.04 10:53:13.122805 [ 27702 ] {} <Debug> Application: Shut down storages.
2021.02.04 10:53:13.123002 [ 27702 ] {} <Debug> Application: Destroyed global context.
2021.02.04 10:53:13.124306 [ 27702 ] {} <Error> Application: Connection refused: 47.106.193.101:7017
2021.02.04 10:53:13.124362 [ 27702 ] {} <Information> Application: shutting down
2021.02.04 10:53:13.124374 [ 27702 ] {} <Debug> Application: Uninitializing subsystem: Logging Subsystem
2021.02.04 10:53:13.124544 [ 27710 ] {} <Trace> BaseDaemon: Received signal -2
2021.02.04 10:53:13.124595 [ 27710 ] {} <Information> BaseDaemon: Stop SignalListener thread

hi team

how to resolve the problem? 

I find clickhouse-server try to connect port 7017 but I can't find any configuration about the port 

ch works before I restart the server, and I don't change the configuration

### PR DESCRIPTION
**Pull request by vitlibar**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category:
- Bug Fix

Changelog entry:
The `MongoDB` table engine now establishes connection only when it's going to read data. `ATTACH TABLE` won't try to connect anymore.

This PR fixes https://github.com/ClickHouse/ClickHouse/issues/20055

### CODE REVIEW

File: src/Storages/StorageMongoDB.cpp [hunk 1]
```diff
@@ -51,6 +50,26 @@ StorageMongoDB::StorageMongoDB(
 }
 
 
+void StorageMongoDB::connectIfNotConnected()
+{
+    std::lock_guard lock{connection_mutex};
+    if (!connection)
+        connection = std::make_shared<Poco::MongoDB::Connection>(host, port);
```
Comment (KochetovNicolai) [inline src/Storages/StorageMongoDB.cpp:57]: I think it is better to release mutex right after connection is initialized.

**Review by KochetovNicolai — APPROVED** [review]

File: src/Storages/StorageMongoDB.cpp (hunk 1 above)
Comment (vitlibar) [reply src/Storages/StorageMongoDB.cpp:57]: It seems there is no point in allowing multiple threads to do the authentication in parallel.

File: src/Storages/StorageMongoDB.h [hunk 2]
```diff
@@ -40,16 +40,19 @@ class StorageMongoDB final : public ext::shared_ptr_helper<StorageMongoDB>, publ
...
+    const std::string host;
+    const short unsigned int port;
+    const std::string database_name;
+    const std::string collection_name;
+    const std::string username;
+    const std::string password;
 
     std::shared_ptr<Poco::MongoDB::Connection> connection;
+    bool authentified = false;
+    std::mutex connection_mutex; /// Protects the variables `connection` and `authentified`.
```
Comment (vitlibar) [inline src/Storages/StorageMongoDB.h:55]: I've added the comment.

File: src/Storages/StorageMongoDB.cpp (hunk 1 above)
Comment (KochetovNicolai) [reply src/Storages/StorageMongoDB.cpp:57]: ok

**Pull request merged**